jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/CreateS3BucketDialog.kt

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.s3
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.DialogWrapper
+import com.intellij.openapi.ui.ValidationInfo
+import org.jetbrains.annotations.TestOnly
+import software.amazon.awssdk.services.s3.S3Client
+import software.aws.toolkits.resources.message
+import java.awt.Component
+import javax.swing.JComponent
+
+class CreateS3BucketDialog(
+    private val project: Project,
+    private val s3Client: S3Client,
+    private val parent: Component? = null
+) : DialogWrapper(project, parent, false, IdeModalityType.PROJECT) {
+
+    private val view = CreateBucketPanel(project)
+
+    var bucket: String? = null
+        private set
+
+    init {
+        title = message("s3.create.bucket.title")
+        setOKButtonText(message("s3.create.bucket.create"))
+
+        init()
+    }
+
+    override fun createCenterPanel(): JComponent? = view.component
+
+    override fun getPreferredFocusedComponent(): JComponent? = view.bucketName
+
+    override fun doValidate(): ValidationInfo? = validateBucketName()?.let { ValidationInfo(it, view.bucketName) }
+
+    override fun doOKAction() {
+        if (okAction.isEnabled) {
+            setOKButtonText(message("s3.create.bucket.in_progress"))
+            isOKActionEnabled = false
+
+            ApplicationManager.getApplication().executeOnPooledThread {
+                try {
+                    createBucket()
+                    ApplicationManager.getApplication().invokeLater({
+                        close(OK_EXIT_CODE)
+                    }, ModalityState.stateForComponent(view.component))
+                } catch (e: Exception) {
+                    setErrorText(e.message)
+                    setOKButtonText(message("s3.create.bucket.create"))
+                    isOKActionEnabled = true
+                }
+            }
+        }
+    }
+
+    private fun bucketName(): String = view.bucketName.text.trim()
+
+    private fun createBucket() {
+        s3Client.createBucket { request -> request.bucket(bucketName()) }
+        bucket = bucketName()
+    }
+
+    @TestOnly
+    fun validateBucketName(): String? = if (bucketName().isEmpty()) message("s3.create.bucket.missing.bucket.name") else null
```

Comment:
Should we do more validation - s3 publishes a pretty good set of things that it'll accept.

ps - why expose this method? can't we just call `doValidate` directly?

Reply:
S3 bucket naming has many restrictions and they are changing overtime. https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html

The SDKs don't do client side validations and I have the same concerns. Ie. if S3 removes some restriction, the old version of the plugin won't allow customer to create bucket.

Reply:
`doValidate` is a `protected` method.